### PR TITLE
Mobile Endpoint Access Permissions

### PR DESCRIPTION
**Pull request by ctsims**
[Trello Card](https://trello.com/c/gfl9nnI7/32-mobile-endpoint-access-permissions).

Introduces a feature flag to allow setting a role permissions which dictates whether users have access to mobile endpoints. This flag allows locking down access to the full data export for "mobile users" who only use Web Apps and otherwise wouldn't have access to the full offline data set.

Relies on: https://github.com/dimagi/formplayer/pull/724 to ensure that formplayer still works properly for users without the permission

##### SUMMARY
Feature flagged for now even though the default permission set would still allow users to access. This should make it easy to remove the flag at a later date.

This PR and 

##### FEATURE FLAG
New Flag: "Require explicit permissions to access mobile app endpoints" - Custom for now. May become default.

### CODE REVIEW

File: corehq/apps/users/templates/users/partials/edit_role_modal.html [hunk 1]
```diff
@@ -583,6 +583,31 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
             </div>
             </div>
           {% endif %}
+	  {% if request|toggle_enabled:"RESTRICT_MOBILE_ACCESS" %}
```
Comment (orangejenny) [inline corehq/apps/users/templates/users/partials/edit_role_modal.html:586]: Nitpick: is this file mixing tabs and spaces? Indentation is weird.

File: corehq/apps/ota/decorators.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from corehq.apps.users.decorators import require_permission
+from corehq.apps.users.models import Permissions
+
+from dimagi.utils.couch.cache.cache_core import get_redis_client
+
+from functools import wraps
+
+from django.http import HttpResponseForbidden
+
+ORIGIN_TOKEN_HEADER = 'HTTP_X_COMMCAREHQ_ORIGIN_TOKEN'
```
Comment (orangejenny) [inline corehq/apps/ota/decorators.py:11]: Is there magic somewhere that's turning `X-CommCareHQ-Origin-Token` into this? I see the pattern elsewhere but don't understand how it works.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/ota/decorators.py (hunk 2 above)
Comment (ctsims) [reply corehq/apps/ota/decorators.py:11]: Yeah, I was also really confused (am still somewhat confused)

I think Django didn't support a clean headers bucket until a very recent version (that we aren't on, I tried to use it), and in the mean time their approach was to pile a bunch of other request metadata into a single dict, and prepend all of the HTTP headers into that bucket with "HTTP_", upcase them, and replace all non[a-zA-Z] chars with `_`. 

Someone who's more familiar with django probably has a fun story about why this definitely made sense at some point and has been a massive painpoint for everyone else for the last 10 years.

File: corehq/apps/users/templates/users/partials/edit_role_modal.html (hunk 1 above)
Comment (ctsims) [reply corehq/apps/users/templates/users/partials/edit_role_modal.html:586]: Ugh, yeah, that's my bad. Thanks for catching. Will clean up before I push

**Comment by ctsims** [comment]
@dannyroberts Do you see any security risks in the Redis keying here? Wanted to make sure you didn't see a path to someone maliciously injecting origin tokens.

**Comment by dannyroberts** [comment]
Yeah it's an important part of our security model that it's impossible for a user of commcare to set an arbitrary key-value pair in redis of their choosing

**Pull request merged**